### PR TITLE
Allow environment variables in #[files] attributes

### LINKED ISSUE #276: Using `${ENV_VAR}` environment variables in `#[files]` possible?
I have a test suite of JavaScript files that I really cannot put in my repo (it's the WPT, which a shallow clone can take a gigabyte). Unfortunately, that means that I would have `#[files()]` that point to files outside my repo, which I also don't like much.

Would it be possible to use an environment variable in the path so people can provide the path to their local copy of those suites?

Users running the rstests would have opted-in using a feature flag. If that environment variable is not set, it'd be okay with me to just panic (with a friendly message of course).

If this is something that could be done and the maintainers are interested, I can make the PR for it. Let me know.

### PR DESCRIPTION
**Pull request by hansl**
The environment variable must be passed to cargo test. Any variable name is valid and would be replaced. If a variable is not defined, the macro will error with the proper error message.

Fixes #276

### CODE REVIEW

**Comment by hansl** [comment]
I changed the test to add a `HELLO_WORLD`. I'm going to add a proper test, but this is just to show an example.

Here is the result when running with the environment variable undeclared:

CleanShot 2024-09-24 at 09 32 37@2x

Here is the same test with a proper value (just having fun doing up and down on path):

CleanShot 2024-09-24 at 09 33 20@2x

Please note that if the environment variable is declared but doesn't go into the proper folder, it is equivalent to specifying the wrong path in `#[files]`. So the tests don't run.

**Review by la10736 — CHANGES_REQUESTED** [review]
Really nice  JOB!!! Thanks for your contribution.

Can you please add some unit tests, a line in change log and add document this new feature in the https://github.com/la10736/rstest/blob/master/rstest/src/lib.rs#L954 and https://github.com/la10736/rstest/blob/a6a8cd9bc1441796a511694f9983243968f4bd51/README.md?plain=1#L233 ?

File: rstest_macros/src/parse/rstest/files.rs [hunk 1]
```diff
@@ -16,11 +16,54 @@ use crate::{
...
+            Err(_) => {
+                return Err(attr.error(&format!(
+                    "Could not find the environment variable {:?}",
+                    var_name
+                )))
+            }
+        };
+
+        // Make sure cargo would rerun if the value of this environment variable changes.
+        println!("cargo::rerun-if-env-changed={var_name}");
```
[the remark this answers is not in the document]
Comment (hansl) [reply rstest_macros/src/parse/rstest/files.rs:49]: It works in my project. Just rerunning `cargo test` without changing the environment variable will not rebuild the project, but changing it will rebuild.

File: rstest_macros/src/parse/rstest/files.rs [hunk 2]
```diff
@@ -213,10 +271,17 @@ impl VisitMut for ValueFilesExtractor {
                     .push(attr.error("Cannot use #[include_dot_files] more than once"))
             })
         }
+        let ignore_missing_env_vars = self.extract_ignore_missing_env_vars(node);
+        if !ignore_missing_env_vars.is_empty() {
+            ignore_missing_env_vars.iter().skip(1).for_each(|attr| {
+                self.errors
+                    .push(attr.error("Cannot use #[ignore_missing_env_vars] more than once"))
+            })
+        }
```
[the remark this answers is not in the document]
Comment (hansl) [reply rstest_macros/src/parse/rstest/files.rs:360]: Done.

File: rstest_macros/src/parse/rstest/files.rs [hunk 3]
```diff
@@ -16,11 +16,54 @@ use crate::{
     utils::attr_is,
 };
 
+/// Replace environment variables in a string.
+fn replace_env_vars(attr: &LitStrAttr, ignore_missing: bool) -> Result<String, syn::Error> {
+    let re =
+        Regex::new(r"\$([a-zA-Z_][a-zA-Z0-9_]*)|\$\{([^}]*)}").expect("Could not build the regex");
```
[the remark this answers is not in the document]
Comment (hansl) [reply rstest_macros/src/parse/rstest/files.rs]: Let me add more of that.

**Comment by hansl** [comment]
@la10736 Thanks for the review.

Added a new `#[base_dir = '...']` to allow to change the base dir for a test. It helps when using absolute paths (which cannot be used for `#[files(...)]`). Added tests for pretty much everything in this PR (checked locally with coverage). Also added a couple more integration tests.

PTAL

**Review by la10736 — APPROVED** [review]
That's an amazing work. Thanks again. 
I would have preferred that formatting was separated by code changing and some more unit tests about parsing: but I really appreciated your work and I should find a compromise with my desiderata :smile: 

I'm not sure if I really like the `base_dir` feature but maybe it can be useful in some context.

**Comment by la10736** [comment]
Just a note about tests and coverage: pass in every lines (100% coverage) doesn't means spot all  issues.... especially on on algorithms where the corner cases are not trivially identified by if branches. Anyway, your code and tests are really great!

File: rstest_macros/src/parse/rstest/files.rs [hunk 4]
```diff
@@ -15,21 +9,114 @@ use crate::{
...
+                (None, _) => {
+                    return Err(attr.error(&format!(
+                        "Could not find the environment variable {:?}",
+                        var_name
+                    )))
+                }
+            };
+
+            // Make sure cargo would rerun if the value of this environment variable changes.
+            println!("cargo::rerun-if-env-changed={var_name}");
```
Comment (la10736) [inline rstest_macros/src/parse/rstest/files.rs:90]: Unfortunately it doesn't seem to work. I tried it this feature for base root and if just change, add or remove this variable the tests are not recompiled. If I add if to a build script it seams to work.

I'll remove it and add some documentation notes.

Comment (hansl) [reply rstest_macros/src/parse/rstest/files.rs:90]: I tested it on my side, let me investigate a bit.
